_posts/2018-10-08-post-javascript-object-cloning.md

```diff
@@ -0,0 +1,90 @@
+image_credit: thuyen 0
+---
+In Javascript, we are always working with different data types to send to the API or display on the UI. Cloning and altering that clone is a very common thing we do every day. Sometimes we only want to alter the newly cloned data with affecting the original data. For example, we can copy an array and alter the new one and everything works out fine.
+
+```javascript
+let originalArray = [1,2,3,4];
+let newArray = originalArray;
+newArray.push(5);
+
+console.log(originalArray);
+[1,2,3,4]
+
+console.log(newArray);
+[1,2,3,4,5]
+```
+
+As you can see we can alter the newArray and push the integer of 5 in without altering the originalArray. This is nothing new and is covered for a basic of Javascript. There can be a basic gotcha when we need to clone objects. For this example, we are going to imagine we have an array of objects from the api. We want to display that data and all filtering is going to be only UI filtering. The idea is to take input from the user, clone the original array we got from the api and apply the filters without altering the original array. This is where deep cloning and shallowing cloning comes in.  Below is way to “clone” this array.
+
+```javascript
+let clientList = [
+  {
+    "firstName": "Chuck",
+    "lastName": "Norris",
+    "address": "1600 Pennsylvania Ave NW",
+    "city": "Washington",
+    "state": "DC",
+    "zip": "20500"
+  },
+  {
+    "firstName": "John",
+    "lastName": "Snow",
+    "address": "",
+    "city": "",
+    "state": "",
+    "zip": ""
+  },
+  {
+    "firstName": "Micky",
+    "lastName": "Mouse",
+    "address": "1313 Disneyland Dr",
+    "city": "Anaheim",
+    "state": "CA",
+    "zip": "92802"
+  }
+]
+```
+
+Now if we create a new array called `filteredList` by just assigned it from `clientList` and alter an object you will see it reflect in both variables.
+
+```javascript
+let filteredList = clientList;
+
+filteredList.[0].firstName= “Billy”;
+
+console.log(filteredList);
+firstName: “Billy”
+console.log(clientList);
+firstName: “Billy”
+```
+
+This is happening because when we created the new array `filteredList` we are only shallow cloning `clientList`. Instead of copying the objects into the new array we are only copying a reference back to the original object. If you alter a reference object it really changes the original object. This can be an issue if we want to keep the integrity of the original api response object. 
+
+## Easy and simple
+
+An easy way to handle this situation is to use `JSON.parse` and `JSON.stringify`. This will do a deep clone of `clientList` and you can alter `filteredList` without effecting `clientList`. 
```

Comment:
```
An easy way to handle this situation is to use `JSON.parse` and `JSON.stringify`. This will do a deep clone of `clientList` and you  can alter `filteredList` without affecting `clientList`.
```
(Just changed effecting to affecting)

It might be nice in here to also just add a sentence explaining _why_ this works. :smile: Since strings are primitives, by changing it to a primitive value, you remove any references so that when you swap it back you have a clean copy to work with.